### PR TITLE
add xds adaptor to config cluster(CDS).

### PR DESCRIPTION
**Pull request by ma642**
**What this PR does**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Review by AlexStocks — APPROVED** [review]

File: go.mod [hunk 1]
```diff
@@ -51,4 +53,5 @@ require (
 	vimagination.zapto.org/memio v0.0.0-20200222190306-588ebc67b97d // indirect
 )
 
-replace github.com/envoyproxy/go-control-plane => github.com/envoyproxy/go-control-plane v0.8.0
+//replace github.com/dubbogo/dubbo-go-pixiu-filter => ../dubbo-go-pixiu-filter
```
Comment (mark4z) [inline go.mod]: clean unused code maybe better

File: pkg/adapter/xds/README.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+### run unit test
+Base some grpc function/method will be patched when run unit test, please set arguments for run test like below:
+```shell
+GOARCH=amd64 go test -gcflags=-l -v -cover .
```
Comment (mark4z) [inline pkg/adapter/xds/README.md:113]: Can you give more explanation?

File: pkg/adapter/xds/apiclient/grpc.go [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package apiclient
+
+import (
+	"context"
+	"dubbo.apache.org/dubbo-go/v3/common/logger"
```
Comment (mark4z) [inline pkg/adapter/xds/apiclient/grpc.go]: Wrong logger, use 	"github.com/apache/dubbo-go-pixiu/pkg/logger"

File: pkg/adapter/xds/apiclient/apiclient.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+ * limitations under the License.
+ */
+
+package apiclient
+
+import (
+	v3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	"github.com/pkg/errors"
+	"google.golang.org/protobuf/proto"
+	"reflect"
```
Comment (ztelur) [inline pkg/adapter/xds/apiclient/apiclient.go]: use imports-formatter -path . -module github.com/apache/dubbo-go-pixiu -bl false to split import block

File: pkg/adapter/xds/apiclient/grpc.go [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+
+	GrpcCluster interface {
+		GetConnect() *grpc.ClientConn
+	}
+	GrpcClusterManager interface {
+		GetGrpcCluster(name string) (GrpcCluster, error)
+	}
+)
+
+func CreateGrpcApiClient(config *model.ApiConfigSource, node *model.Node,
```
Comment (ztelur) [inline pkg/adapter/xds/apiclient/grpc.go:70]: add comment for public struct or function

File: pkg/adapter/xds/apiclient/grpc.go [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+		Id:            g.node.Id,
+		Cluster:       g.node.Cluster,
+		UserAgentName: xdsAgentName,
+	}
+}
+
+func (g *GrpcApiClient) Delta() (chan *DeltaResources, error) {
+	outputCh := make(chan *DeltaResources)
+	return outputCh, g.runDelta(outputCh)
+	//return outputCh, g.runStream(outputCh)
```
Comment (ztelur) [inline pkg/adapter/xds/apiclient/grpc.go]: delete useless line

File: pkg/adapter/xds/apiclient/grpc.go [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+
+				continue //backoff
+			}
+			return //success
+		}
+	}
+	backoff()
+	//get message
+	go func() {
+		for { // delta response backoff.
```
Comment (ztelur) [inline pkg/adapter/xds/apiclient/grpc.go:168]: should handle exitCh close event?

File: pkg/adapter/xds/apiclient/grpc.go [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+	}
+	logger.Infof("get xDS message nonce, %s", resp.Nonce)
+	for _, res := range resp.RemovedResources {
+		logger.Infof("remove resource found ", res)
+		resources.RemovedResource = append(resources.RemovedResource, res)
+	}
+
+	for _, res := range resp.Resources {
+		logger.Infof("new resource found %s version=%s", res.Name, res.Version)
+		g.xdsState.deltaVersion[res.Name] = res.Version
```
Comment (ztelur) [inline pkg/adapter/xds/apiclient/grpc.go:207]: 这块不了解，想请教一下。这个版本信息不需要向下传递嘛？用作后续 resources 的处理，比如说发现本地的版本高，就抛弃掉刚接收到的低版本的 resource

File: pkg/adapter/xds/apiclient/grpc.go [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+		ResponseNonce:            g.xdsState.nonce,
+		ErrorDetail:              nil,
+	})
+	if err != nil {
+		return nil, errors.Wrapf(err, "can not send delta discovery request")
+	}
+	return delta, nil
+}
+
+func (g *GrpcApiClient) runStream(output chan<- *DeltaResources) error {
```
Comment (ztelur) [inline pkg/adapter/xds/apiclient/grpc.go]: is this function unused ?

File: pkg/adapter/xds/cds.go [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+func (c *CdsManager) removeCluster(clusterNames []string) {
+	server.GetClusterManager().RemoveCluster(clusterNames)
+}
+
+func (c *CdsManager) setupCluster(clusters []*xdspb.Cluster) error {
+	clusterMgt := server.GetClusterManager()
+	laterApplies := make([]func() error, 0, len(clusters))
+	toRemoveHash := make(map[string]struct{}, len(clusters))
+
+	for _, _cluster := range clusters {
```
Comment (ztelur) [inline pkg/adapter/xds/cds.go]: should replace `clusters ` with `clusterMgt.getAllCluster` to figure wait-deleted clusters.

File: pkg/adapter/xds/cds.go [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+		Address:      endpoint.Address.Address,
+		Port:         int(endpoint.Address.Port),
+		ResolverName: endpoint.Address.ResolverName,
+		Domains:      endpoint.Address.Domains,
+		CertsDir:     endpoint.Address.CertsDir,
+	}
+}
+
+func (c *CdsManager) makeHealthChecks(checks []*xdspb.HealthCheck) (result []model.HealthCheck) {
+	//todo implement me after fix model.HealthCheck type define
```
Comment (ztelur) [inline pkg/adapter/xds/cds.go:184]: just set a default value ?

File: pkg/adapter/xds/xds.go [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+		//	if creds, err = xdscreds.NewClientCredentials(xdscreds.ClientOptions{FallbackCreds: insecure.NewCredentials()}); err != nil {
+		//		log.Fatalf("failed to create client-side xDS credentials: %v", err)
+		//	}
+		//}
+		if len(g.config.Endpoints) == 0 {
+			panic("expect endpoint.")
+		}
+		address := g.config.Endpoints[0].Address.Address
+		port := g.config.PickOneEndpoint().Address.Port
+		endpoint := fmt.Sprintf("%s:%d", address, port)
```
Comment (ztelur) [inline pkg/adapter/xds/xds.go]: use `Address.GetAddress` function.

File: pkg/adapter/xds/xds.go [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+	if dm.GetLds() != nil {
+		a.lds = &LdsManager{DiscoverApi: a.createApiManager(dm.GetLds(), dm.GetNode(), xds.ListenerType)}
+		if err := a.lds.Fetch(); err != nil {
+			logger.Errorf("can not fetch lds err is %+v", err)
+		}
+	}
+	// catch the ongoing cds config change.
+	if dm.GetCds() != nil {
+		a.cds = &CdsManager{DiscoverApi: a.createApiManager(dm.GetCds(), dm.GetNode(), xds.ClusterType)}
+		if err := a.cds.Delta(); err != nil {
```
Comment (ztelur) [inline pkg/adapter/xds/xds.go:204]: should csd call `Fetch` firstly and then call `Delta`?

**Comment by ztelur** [comment]
Thanks for your excellent work.

**Review by ztelur — APPROVED** [review]

File: samples/xds/local-control-panel/readme.md [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+```
+
+2. run pixiu 
+```
+dubbo-go-pixiu > pixiu gateway start -c ./samples/xds/pixiu/conf.yaml -g test/configs/log.yml
+```
+
+3. check result
+```shell
+curl -v  'localhost:8889/get'
```
Comment (mark4z) [inline samples/xds/local-control-panel/readme.md]: port should be 8888

File: pkg/adapter/xds/lds.go [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+ */
+
+package xds
+
+import (
+	"encoding/json"
+)
+
+import (
+	model2 "github.com/dubbogo/dubbo-go-pixiu-filter/pkg/xds/model"
```
Comment (mark4z) [inline pkg/adapter/xds/lds.go]: xdsmodel or any name?

File: pkg/adapter/xds/lds.go [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+
+// Fetch overwrite DiscoverApi.Fetch.
+func (l *LdsManager) Fetch() error {
+	r, err := l.DiscoverApi.Fetch("") //todo use local version
+	if err != nil {
+		return err
+	}
+	listeners := make([]*model2.Listener, 0, len(r))
+	for _, one := range r {
+		_listener := &model2.PixiuExtensionListeners{}
```
Comment (mark4z) [inline pkg/adapter/xds/lds.go]: why "_listener" not "listener"

**Comment by mark4z** [comment]
XDS as a adapter， sounds weird

**Review by mark4z — APPROVED** [review]

**Pull request merged**